control-plane-operator/controllers/hostedcontrolplane/autoscaler/reconcile.go

```diff
@@ -1,46 +1,106 @@
 	"github.com/openshift/hypershift/support/config"
-	"github.com/openshift/hypershift/support/upsert"
+	component "github.com/openshift/hypershift/support/controlplane-component"
 	"github.com/openshift/hypershift/support/util"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	rbacv1 "k8s.io/api/rbac/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
-	k8sutilspointer "k8s.io/utils/pointer"
+	"k8s.io/utils/ptr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
 	autoscalerName = "cluster-autoscaler"
+
+	ImageStreamAutoscalerImage = "cluster-autoscaler"
 )
 
-func ReconcileAutoscalerDeployment(deployment *appsv1.Deployment, hcp *hyperv1.HostedControlPlane, sa *corev1.ServiceAccount, kubeConfigSecret *corev1.Secret, options hyperv1.ClusterAutoscaling, clusterAutoscalerImage, availabilityProberImage string, setDefaultSecurityContext bool, ownerRef config.OwnerRef) error {
-	ownerRef.ApplyTo(deployment)
+var _ component.DeploymentReconciler = &AutoscalerReconciler{}
+
+type AutoscalerReconciler struct {
+}
+
+func NewComponent() component.ControlPlaneComponent {
+	return &component.ControlPlaneWorkload{
+		DeploymentReconciler:     &AutoscalerReconciler{},
+		RBACReconciler:           component.NewRBACReconciler(autoscalerRoleRules()),
+		Predicate:                Predicate,
+		NeedsManagementKASAccess: true,
+	}
+}
+
+// Name implements controlplanecomponent.DeploymentReconciler.
+func (a *AutoscalerReconciler) Name() string {
+	return autoscalerName
+}
+
+// Volumes implements controlplanecomponent.DeploymentReconciler.
+func (a *AutoscalerReconciler) Volumes(cpContext component.ControlPlaneContext) component.Volumes {
+	volumeSource := component.SecretVolumeSource(manifests.KASServiceCAPIKubeconfigSecret(cpContext.HCP.Namespace, cpContext.HCP.Spec.InfraID).Name)
+	volumeSource.Secret.Items = []corev1.KeyToPath{
+		{
+			// TODO: should the key be published on status?
+			Key:  "value",
+			Path: "target-kubeconfig",
+		},
+	}
+
+	return component.Volumes{
+		"target-kubeconfig": component.Volume{
+			Source: volumeSource,
+			Mounts: map[string]string{
+				autoscalerName: "/mnt/kubeconfig",
+			},
+		},
+	}
+}
+
+func Predicate(cpContext component.ControlPlaneContext) (bool, error) {
+	hcp := cpContext.HCP
+
+	// Disable cluster-autoscaler component if DisableMachineManagement label is set.
+	if _, exists := hcp.Annotations[hyperv1.DisableMachineManagement]; exists {
+		return false, nil
+	}
+
+	// The deployment depends on the kubeconfig being reported.
+	if hcp.Status.KubeConfig == nil {
+		return false, nil
+	}
+	// Resolve the kubeconfig secret for CAPI which the autoscaler is deployed alongside of.
+	capiKubeConfigSecret := manifests.KASServiceCAPIKubeconfigSecret(hcp.Namespace, hcp.Spec.InfraID)
+	err := cpContext.Client.Get(cpContext, client.ObjectKeyFromObject(capiKubeConfigSecret), capiKubeConfigSecret)
+	if err != nil {
```

Comment:
should this return an error if it's just not found?

Reply:
fixed